projects/28_gltf/main.cpp

```diff
@@ -0,0 +1,837 @@
+
+        PPX_CHECKED_CALL(GetDevice()->CreateDescriptorSetLayout(&layoutCreateInfo, &mSetLayout));
+    }
+
+    Timer timer;
+    timer.Start();
+    LoadScene(
+        "basic/models/altimeter/altimeter.gltf",
+        GetDevice(),
+        GetSwapchain(),
+        GetGraphicsQueue(),
+        mDescriptorPool,
+        &mTextureCache,
+        &mObjects,
+        &mPrimitives,
+        &mMaterials);
+    printf("geometry loaded in %lfs\n", timer.SecondsSinceStart());
+
+    // Per frame data
+    {
+        PerFrame frame = {};
+
+        PPX_CHECKED_CALL(GetGraphicsQueue()->CreateCommandBuffer(&frame.cmd));
+
+        grfx::SemaphoreCreateInfo semaCreateInfo = {};
+        PPX_CHECKED_CALL(GetDevice()->CreateSemaphore(&semaCreateInfo, &frame.imageAcquiredSemaphore));
+
+        grfx::FenceCreateInfo fenceCreateInfo = {};
+        PPX_CHECKED_CALL(GetDevice()->CreateFence(&fenceCreateInfo, &frame.imageAcquiredFence));
+
+        PPX_CHECKED_CALL(GetDevice()->CreateSemaphore(&semaCreateInfo, &frame.renderCompleteSemaphore));
+
+        fenceCreateInfo = {true}; // Create signaled
+        PPX_CHECKED_CALL(GetDevice()->CreateFence(&fenceCreateInfo, &frame.renderCompleteFence));
+
+        mPerFrame.push_back(frame);
+    }
+
+    GetDevice()->DestroyShaderModule(mVertexShader);
+    GetDevice()->DestroyShaderModule(mPixelShader);
+}
+
+void ProjApp::Render()
+{
+    PerFrame&          frame      = mPerFrame[0];
+    grfx::SwapchainPtr swapchain  = GetSwapchain();
+    uint32_t           imageIndex = UINT32_MAX;
+    PPX_CHECKED_CALL(swapchain->AcquireNextImage(UINT64_MAX, frame.imageAcquiredSemaphore, frame.imageAcquiredFence, &imageIndex));
+    // Wait for and reset image acquired fence
+    PPX_CHECKED_CALL(frame.imageAcquiredFence->WaitAndReset());
+    // Wait for and reset render complete fence
+    PPX_CHECKED_CALL(frame.renderCompleteFence->WaitAndReset());
+
+    // Update camera(s)
+    mCamera.LookAt(float3(19, 1, -3), float3(0, 0, 0));
+
+    // Update uniform buffers
+    for (auto& object : mObjects) {
+        // Draw uniform buffers
+        struct Scene
+        {
+            float4x4 modelMatrix;                // Transforms object space to world space
+            float4   ambient;                    // Object's ambient intensity
+            float4x4 cameraViewProjectionMatrix; // Camera's view projection matrix
+            float4   lightPosition;              // Light's position
+            float4   eyePosition;
+        };
+
+        Scene scene                      = {};
+        scene.modelMatrix                = object.model;
+        scene.ambient                    = float4(0.3f);
+        scene.cameraViewProjectionMatrix = mCamera.GetViewProjectionMatrix();
+        scene.lightPosition              = float4(mLightPosition, 0);
+        scene.eyePosition                = glm::float4(mCamera.GetEyePosition(), 0.f);
+
+        object.pUniformBuffer->CopyFromSource(sizeof(scene), &scene);
+    }
+
+    {
```

Comment:
Do we need to update the descriptors at every frame? I don't see anything that changes across frames.

Reply:
We don't for now (no camera movement), but we should add an orbital camera right after this is implemented.